tutorials/examples/mRNA_design/MFE_calculator.py

```diff
@@ -0,0 +1,559 @@
+            n = j1 - i
+            # The energy (penalty) is in the hairpin loop table:
+            e = self._hairpin_fe
+
+        elif nb == 1:
+            # (i,j) closes the internal or bulge loop [i1,j1].
+            # The number of unpaired nucleotides outside the branch is n.
+            p, q = b[0]
+            n = j1 - i1 - (q - p)
+            if p == i1 or q == j1:
+                # The energy (penalty) is in the bulge loop table:
+                e = self._bulge_fe
+            else:
+                # The energy (penalty) is in the internal loop table:
+                e = self._internal_fe
+            # The branch (p,q) contributes energy:
+            v1 += V[p, q]
+
+        else:
+            # (i,j) closes the multi-branch loop [i1,j1].
+            # The number of unpaired nucleotides outside any branch is n.
+            n = j - i1 - nb
+            for p, q in b:
+                # Nucleotides under the branch (p,q) do not count in n.
+                n -= q - p
+                # The branch (p,q) contributes energy:
+                v1 += V[p, q]
+            # The energy (penalty) is in the internal loop table:
+            e = self._internal_fe
+
+        # Consider as loop closure only if not stacking.
+        if n > 0 or nb != 1:
+            v2 = e[n if n < len(e) else -1]
+            vmin = v1 + v2
+            mode = 0  # interior is open (if loop wins)
+
+        # Hypothesise that (i,j) stacks on helix [i1,j1].
+        if (s := tuple_as_pair_int(si, sj)) >= 0:
+            # (i1,j1) can in fact pair. Suppose [i1,j1] to be closed.
+            v1 = self._stack_fe[r, s]
+            v2 = V[i1, j1] if j1 > i1 else 0.0
+            if (v := v1 + v2) < vmin:
+                vmin = v
+                mode = 1  # interior is closed (stacking wins)
+
+        return vmin, mode
+
+    def open_energy(self, soln: RNASolution, i: int, j: int) -> tuple[float, int]:
+        """Minimum energy of [i, j] without pairing (i,j).
+
+        The method seeks a lowest energy for [i,j], considering j unpaired
+        or paired with any interior nucleotide k, where i<k<j.
+        """
+        # Make the working arrays local
+        K, U, V = soln.get_arrays()
+
+        # Consider several hypotheses about the addition of j
+        # to the range [i,j-1] without pairing them. In each hypothesis
+        # where umin is lowered, kmin remembers the pairing of j.
+        j1 = j - 1
+
+        # Hypothesise that j dangles right from structure [i,j-1]
+        u0 = U[i, j1]
+        u1 = V[i, j1]
+        umin = min(u0, u1)
+        kmin = -1
+
+        # Hypothesise that j pairs with k and [i,k-1] is optimally folded.
+        for k in range(i + 1, j):
+            u0 = U[i, k - 1]
+            u1 = V[i, k - 1]
+            u2 = V[k, j]
+            if (u := min(u0, u1) + u2) < umin:
+                umin = u
+                kmin = k
+
+        return umin, kmin
+
+
+class RNAFolder(RNAFolderBase2):
```

Comment:
OK So I haven't followed entirely this logic but to my eye it looks like all we need are two classes:

`RNASolution`, `RNAFolder`.

Would this work? It seems most of the class base factorization isn't used anywhere else?